accepted/cranelift-egraph.md

```diff
@@ -0,0 +1,1188 @@
+will actually be used for optimization rewrites.
+
+### Recap: E-graphs
+
+An e-graph, or "equivalence graph", is a representation of a set of
+value nodes, each of which may use other nodes as arguments, that
+allows *equivalences* between these nodes to be recorded. It
+efficiently represents *all possible expressions* for a value, given
+these equivalences. It has normalizing properties: "adding" the same
+expression to an e-graph multiple times will result in exactly the
+same ID.
+
+An e-graph consists of a set of *classes* (or "eclasses"), each of
+which has a set of equivalent *nodes* (or "enodes"). An eclass has an
+identity (in our implementation based on the `egg` library, simply a
+numeric index). Enodes, however, do not have an identity or name. A
+"value" in the program is represented by an eclass ID.
+
+Aside from a map of eclass ID to eclass data (with list of nodes), an
+e-graph contains (i) a hashmap of node contents to eclass ID, to allow
+for deduplication ("hash-consing") when adding new nodes; and (ii) a
+[union-find](https://en.wikipedia.org/wiki/Disjoint-set_data_structure)
+data structure to enable efficient merging of eclasses and looking up
+canonical eclass IDs post-merge.
+
+The algorithmic and implementation details of e-graphs are beyond the
+scope of this RFC, but the [egg
+paper](https://dl.acm.org/doi/10.1145/3434304)[^1] gives a good
+overview, including some novel contributions of that paper to make
+e-graph updates asymptotically more efficient.
+
+[^1]: M Willsey, C Nandi, Y R Wang, O Flatt, Z Tatlock, P
+      Panchekha. "egg: Fast and Flexible Equality Saturation." In
+      POPL 2021. https://dl.acm.org/doi/10.1145/3434304
+      
+The key relevant facts here are:
+
+- An e-graph lets us build expressions out of *nodes*, each of which
+  contains references to other nodes. This is very much like a
+  sea-of-nodes IR.
+  
+- An e-graph lets us equate one node with another, such that when we
+  examine the value (eclass ID), we can see both nodes and eventually
+  choose either for code generation.
+  
+The first fact is relevant to how we express a function body in an
+e-graph, and we will explore that question in this section. The second
+fact is relevant to how we use an e-graph to progressively rewrite the
+function, and we will explore that in the next section.
+
+### Control Flow in E-graphs
+
+Traditionally, e-graphs have been used to represent expression nodes
+in a control-flow-free representation (or at least, we have not been
+able to find an example of a system that embeds control flow into an
+e-graph). Cranelift function bodies expressed in CLIF have basic
+blocks of instructions in a control-flow graph (CFG); we thus need
+some way to embed the information in the CFG within the e-graph, so
+that when we generate code from the optimized function body within the
+e-graph, we retain equivalent control flow.
+
+There have been numerous *graph-based IRs* of other forms that *do*
```

Comment:
I've not had a chance to read this paper yet [1], but do you know if PEG's are another graph-based IR in this category?

[1] https://rosstate.org/publications/eqsat/

Reply:
Thanks for the link -- I wasn't aware of PEGs previously, but they look to be in the same family as the other approaches! The theta-nodes for loops (see e.g. Fig. 2 in the POPL 2009 paper) indicate to me that these are region-based, like VSDGs, so the same discussion points apply (vs. the retained-CFG-skeleton approach here). I see also in the description that they compute each node at most once, and as a result do partial-redundancy elimination, whereas we take an approach trying to preserve our existing GVN semantics and only compute a value when we know it will be used, which is generally safer from an optimization-always-improves-perf point of view.

Reply:
I haven't checked whether this applies to Program Expression Graphs specifically, but similar dataflow-based IRs like the [Value Dependence Graph](https://homes.cs.washington.edu/~mernst/pubs/vdg-popl94.pdf) don't preserve non-termination. A loop that doesn't terminate might not define any values used later and would then be removed, "because CFG values that are not demanded do not appear." Also, the "Equality Saturation" paper you cited has a whole side thing (section 6.1) about handling side-effects like heap access, because that's another challenge for purely dataflow IRs.

This is what distinguishes the Value State Dependence Graph and the newer Regionalized VSDG. Treating state and termination dependencies as additional pseudo-dataflow edges means any topological sort of the graph is a valid serialization.

I assume Chris' approach preserves termination and side effects (at the cost of missing some optimization opportunities that might be safe) but I haven't looked at it carefully enough yet to understand how it does that. From what I've heard so far I suspect it's different than anything I've seen in the literature.

Reply:
Yes, the approach proposed in this RFC preserves both termination and side-effects in a somewhat simple way: by just keeping that part of the CFG around. Or more precisely, it maintains a "skeleton", which is a list of eclass IDs in each basic block of side-effecting ops in original order. This is then used during the scoped elaboration to schedule the pure ops back into the spaces between them. One nice property of this, also, is that we stay close enough to the original CFG that we could actually do the roundtrip in a way that never has to move instruction data (memory traffic is expensive!) except when a node is truly duplicated: the skeleton's instructions are stored in the original `ir::Function`, and the original copy of each e-node remains in the `ir::Function` as well.